sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala

```diff
@@ -30,8 +34,128 @@ case class GpuJsonToStructs(
+      }
+    }
+  }
+
+  private def processFieldNames(names: Seq[(String, DataType)]): Seq[(String, DataType)] = {
+    val existingNames = Set[String]()
+    // for duplicated field names, only keep the one with the largest index
+    names.foldRight(Seq[(String, DataType)]())((elem, acc) => {
+      val (name, dtype) = elem
+      if (existingNames(name)) (null, dtype)+:acc else {existingNames += name; (name, dtype)+:acc}})
+  }
+
   override protected def doColumnar(input: GpuColumnVector): cudf.ColumnVector = {
-    MapUtils.extractRawMapFromJsonString(input.getBase)
+    schema match {
+      case _: MapType =>
+        MapUtils.extractRawMapFromJsonString(input.getBase)
+      case struct: StructType => {
+        // We cannot handle all corner cases with this right now. The parser just isn't
+        // good enough, but we will try to handle a few common ones.
+        val numRows = input.getRowCount.toInt
+
+        // Step 1: verify and preprocess the data to clean it up and normalize a few things
+        // Step 2: Concat the data into a single buffer
+        val (isNullOrEmpty, combined) = cleanAndConcat(input.getBase)
+        withResource(isNullOrEmpty) { isNullOrEmpty =>
+          // Step 3: copy the data back to the host so we can parse it.
+          val combinedHost = withResource(combined) { combined =>
+            combined.copyToHost()
+          }
+          // Step 4: Have cudf parse the JSON data
+          val (names, rawTable) = withResource(combinedHost) { combinedHost =>
+            val data = combinedHost.getData
+            val start = combinedHost.getStartListOffset(0)
+            val end = combinedHost.getEndListOffset(0)
+            val length = end - start
+
+            withResource(cudf.Table.readJSON(cudf.JSONOptions.DEFAULT, data, start,
+              length)) { tableWithMeta =>
+              val names = tableWithMeta.getColumnNames
+              (names, tableWithMeta.releaseTable())
+            }
+          }
+
+          // process duplicated field names in input struct schema
+          val fieldNames = processFieldNames(struct.fields.map { field => 
+              (field.name, field.dataType)})
+
+          withResource(rawTable) { rawTable =>
+            // Step 5: verify that the data looks correct
+            if (rawTable.getRowCount != numRows) {
+              throw new IllegalStateException("The input data didn't parse correctly and we read " +
+                  s"a different number of rows than was expected. Expected $numRows, " +
+                  s"but got ${rawTable.getRowCount}")
+            }
+
+            // Step 6: get the data based on input struct schema
+            val columns = fieldNames.safeMap { case (name, dtype) =>
+              val i = names.indexOf(name)
+              if (i == -1) {
+                GpuColumnVector.columnVectorFromNull(numRows, dtype)
+              } else {
+                rawTable.getColumn(i).castTo(GpuColumnVector.getRapidsType(dtype))
```

Comment:
Please use GpuCast instead of castTo. IT does a lot more checking to make sure that we parse things correctly/etc.

Reply:
To add to this point, should add integration tests to test any failure conditions that are a result of casting. For example, if you pass in JSON that has arbitrary large number that you specify in your schema to use IntegerType, the same overflow exception that Spark would throw should be checked for.

Reply:
I updated the implementation to use `GpuCast.doCast`. I still observed incompatible CPU vs GPU behavior with arbitrarily large numbers in input JSON strings with IntegerType schema: CPU implementation returns null, and GPU implementation casts the large numbers to integers. I documented this observation in the `compatibility.md` file. I have not yet found a way to work around this while not breaking current existing tests. If it is important to be compatible, I will investigate more into this. Thanks!